builder/llpcBuilderImplArith.cpp

```diff
@@ -36,3 +36,168 @@
+// Create calculation of 2D texture coordinates that would be used for accessing the selected cube map face for
+// the given cube map texture coordinates. Returns <2 x float>.
+Value* BuilderImplArith::CreateCubeFaceCoord(
+    Value*        pCoord,     // [in] Input coordinate <3 x float>
+    const Twine&  instName)   // [in] Name to give instruction(s)
+{
+    Value* pCoordX = CreateExtractElement(pCoord, uint64_t(0));
+    Value* pCoordY = CreateExtractElement(pCoord, 1);
+    Value* pCoordZ = CreateExtractElement(pCoord, 2);
+    Value* pCubeMa = CreateIntrinsic(Intrinsic::amdgcn_cubema, {}, { pCoordX, pCoordY, pCoordZ }, nullptr);
+    Value* pRecipMa = CreateFDiv(ConstantFP::get(getFloatTy(), 1.0), pCubeMa);
+    Value* pCubeSc = CreateIntrinsic(Intrinsic::amdgcn_cubesc, {}, { pCoordX, pCoordY, pCoordZ }, nullptr);
+    Value* pScDivMa = CreateFMul(pRecipMa, pCubeSc);
+    Value* pResultX = CreateFAdd(pScDivMa, ConstantFP::get(getFloatTy(), 0.5));
+    Value* pCubeTc = CreateIntrinsic(Intrinsic::amdgcn_cubetc, {}, { pCoordX, pCoordY, pCoordZ }, nullptr);
+    Value* pTcDivMa = CreateFMul(pRecipMa, pCubeTc);
+    Value* pResultY = CreateFAdd(pTcDivMa, ConstantFP::get(getFloatTy(), 0.5));
+    Value* pResult = CreateInsertElement(UndefValue::get(VectorType::get(getFloatTy(), 2)), pResultX, uint64_t(0));
+    pResult = CreateInsertElement(pResult, pResultY, 1, instName);
+    return pResult;
+}
+
+// =====================================================================================================================
+// Create calculation of the index of the cube map face that would be accessed by a texture lookup function for
+// the given cube map texture coordinates. Returns a single float with value:
+//  0.0 = the cube map face facing the positive X direction
+//  1.0 = the cube map face facing the negative X direction
+//  2.0 = the cube map face facing the positive Y direction
+//  3.0 = the cube map face facing the negative Y direction
+//  4.0 = the cube map face facing the positive Z direction
+//  5.0 = the cube map face facing the negative Z direction
+Value* BuilderImplArith::CreateCubeFaceIndex(
+    Value*        pCoord,     // [in] Input coordinate <3 x float>
+    const Twine&  instName)   // [in] Name to give instruction(s)
+{
+    Value* pCoordX = CreateExtractElement(pCoord, uint64_t(0));
+    Value* pCoordY = CreateExtractElement(pCoord, 1);
+    Value* pCoordZ = CreateExtractElement(pCoord, 2);
+    return CreateIntrinsic(Intrinsic::amdgcn_cubeid, {}, { pCoordX, pCoordY, pCoordZ }, nullptr, instName);
+}
+
+// =====================================================================================================================
+// Create quantize operation: truncates FP (or vector) value to a value that is representable by the target type.
+// Currently only implemented for quantizing float to half.
+Value* BuilderImplArith::CreateQuantize(
```

Comment:
please keep name aligned with SPIRV op code (CreateQuantizeToFp16)

Reply:
I deliberately gave it a more general name so that in the future, if needed, we could extend its functionality to do different quantization (e.g. double to float). Do you really want to make the name completely specific?

Reply:
I don't think we need quantize double to float in short time. fp16 is the only one special case.
for the conversion between FP32 and FP64, we already have float control extension, it uses the rounding modes to control the behavior.